README.rst
OctoPi
======

.. image:: https://raw.githubusercontent.com/guysoft/OctoPi/devel/media/OctoPi.png
.. :scale: 50 %
.. :alt: OctoPi logo

A `Raspberry Pi <http://www.raspberrypi.org/>`_ distribution for 3d printers. It includes the `OctoPrint <http://octoprint.org>`_ host software for 3d printers out of the box and `mjpg-streamer with RaspiCam support <https://github.com/jacksonliam/mjpg-streamer>`_ for live viewing of prints and timelapse video creation. OctoPi also includes `OctoPiPanel <https://github.com/jonaslorander/OctoPiPanel>`_, which is an LCD display app that works with OctoPrint, and scripts to configure supported display.s
This repository contains the source script to generate the distribution out of an existing `Raspbian <http://www.raspbian.org/>`_ distro image.

Where to get it?
----------------

Download the latest stable build via this button:

.. image:: https://i.imgur.com/NvUOGfS.png
   :target: https://octopi.octoprint.org/latest

Official mirror is `here <http://docstech.net/OctoPiMirror/>`_

Nightly builds are available `here <http://gnethomelinux.com/OctoPi/nightly/>`_

How to use it?
--------------

#. Unzip the image and install it to an sd card `like any other Raspberry Pi image <https://www.raspberrypi.org/documentation/installation/installing-images/README.md>`_
#. Configure your WiFi by editing ``octopi-network.txt`` on the root of the flashed card when using it like a thumb drive
#. Boot the Pi from the card
#. Log into your Pi via SSH (it is located at ``octopi.local`` `if your computer supports bonjour <https://learn.adafruit.com/bonjour-zeroconf-networking-for-windows-and-linux/overview>`_ or the IP address assigned by your router), default username is "pi", default password is "raspberry", change the password using the ``passwd`` command and expand the filesystem of the SD card through the corresponding option when running ``sudo raspi-config``.

OctoPrint is located at `http://octopi.local <http://octopi.local>`_ and also at `https://octopi.local <https://octopi.local>`_. Since the SSL certificate is self signed (and generated upon first boot), you will get a certificate warning at the latter location, please ignore it.

If a USB webcam or the Raspberry Pi camera is detected, MJPG-streamer will be started automatically as webcam server. OctoPrint on OctoPi ships with correctly configured stream and snapshot URLs pointing at it. If necessary, you can reach it under `http://octopi.local/webcam/?action=stream <octopi.local/webcam/?action=stream>`_ and SSL respectively, or directly on its configured port 8080: `http://octopi.local:8080/?action=stream <octopi.local:8080/?action=stream>`_.

CuraEngine is installed and OctoPrint ships pre-configured with the correct path to utilize it for on-board-slicing. Just import a Cura Slicing Profile in OctoPrint's settings and start slicing directly on your Pi.

Features
--------

* `OctoPrint <http://octoprint.org>`_ host software for 3d printers out of the box
* `Raspbian <http://www.raspbian.org/>`_ tweaked for maximum preformance for printing out of the box
* `mjpg-streamer with RaspiCam support <https://github.com/jacksonliam/mjpg-streamer>`_ for live viewing of prints and timelapse video creation.
* `CuraEngine <https://github.com/Ultimaker/CuraEngine>`_ pre-installed for slicing directly on the Raspberry Pi
* `OctoPiPanel <https://github.com/jonaslorander/OctoPiPanel>`_, which is an LCD display app that works with OctoPrint
* Configuration scripts for various LCD displays

Developing
----------

Requirements
~~~~~~~~~~~~

#. `qemu-arm-static <http://packages.debian.org/sid/qemu-user-static>`_
#. Downloaded `Raspbian <http://www.raspbian.org/>`_ image.
#. root privileges for chroot
#. Bash
#. git
#. realpath
#. sudo (the script itself calls it, running as root without sudo won't work)

Build OctoPi From within OctoPi / Raspbian / Debian / Ubuntu
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

OctoPi can be built from Debian, Ubuntu, Raspbian, or even OctoPi.
Build requires about 2.5 GB of free space available.
You can build it by issuing the following commands::

    sudo apt-get install gawk util-linux realpath qemu-user-static git
    
    git clone https://github.com/guysoft/OctoPi.git
    cd OctoPi/src/image
    curl -J -O -L  http://downloads.raspberrypi.org/raspbian_latest
    cd ..
    sudo modprobe loop
    sudo bash -x ./build
    
Building OctoPi Variants
~~~~~~~~~~~~~~~~~~~~~~~~

OctoPi supports building variants, which are builds with changes from the main release build. An example and other variants are available in the folder ``src/variants/example``.

To build a variant use::

    sudo bash -x ./build [Variant]
    
Configuration Lpfrg Variant
~~~~~~~~~~~~~~~~~~~~~~~~~~~

1. Place a textfile with the preferred default Pi password in variants/lpfrg/LPFRGPASSWD
2. Place the cloud configuration yaml in variants/lpfrg/filesystem/home/pi/.octoprint/data/lui/cloud.yaml
    
Building Using Vagrant
~~~~~~~~~~~~~~~~~~~~~~
There is a vagrant machine configuration to let build OctoPi in case your build environment behaves differently. Unless you do extra configuration, vagrant must run as root to have nfs folder sync working.

To use it::

    sudo apt-get install vagrant nfs-kernel-server
    sudo vagrant plugin install vagrant-nfs_guest
    sudo modprobe nfs
    cd OctoPi/src/vagrant
    sudo vagrant up

After provisioning the machine, its also possible to run a nightly build which updates from devel using::

    cd OctoPi/src/vagrant
    run_vagrant_build.sh
    
To build a variant on the machine simply run::

    cd FullPageOS/src/vagrant
    run_vagrant_build.sh [Variant]
    

Usage
~~~~~

#. If needed, override existing config settings by creating a new file ``src/config.local``. You can override all settings found in ``src/config``. If you need to override the path to the Raspbian image to use for building OctoPi, override the path to be used in ``ZIP_IMG``. By default the most recent file matching ``*-raspbian.zip`` found in ``src/image`` will be used.
#. Run ``src/build`` as root.
#. The final image will be created at the ``src/workspace``

Code contribution would be appreciated!


CHANGELOG.md
# Changelog 

Changelog for OctoPi for Lpfrg

## 1.2.0

- Added exFAT support
- Flush writes to inserted media faster
- Tlsdate is replaced by htpdate
- Update of LpfrgScreensaver to retrieve model from firmware information
- Command aliases for common OctoPrint unix commands
- Rasbian 2017-04-10 image (kernel 4.4)
- New "devel" flavor to preconfigure virtual printer, disable root readonly, and landscape display orientation
- Several LSB scripts are now migrated to systemd
- Less reboots on initial startup
- Enabled SPI
- Includes OctoPrint-rgbStatus
- No more pink line on WaveShare displays
- NetworkManager state on home partition to maintain state on reboot
- Updated Leapfrog branding

## 1.1.0 

- Initial release of OctoPi for Lpfrg

src/variants/lpfrg/filesystem/root/home/pi/scripts/poweroff.py
#!/usr/bin/env python
import RPi.GPIO as GPIO
import time

COMMAND = "/usr/bin/sudo /sbin/shutdown -h now"
LED_PIN = 22 
PWR_PIN = 11 # Toggles power supply
BUTTON_PIN = 16

BOUNCETIME = 100 # minimal press interval in ms
RESETPRESSTIME = 15 # maximum press duration for regular power off

class PowerButtonHandler:
	def __init__(self):
		self.started = None
		self.ended = None

	def press_event(self, e):
		if GPIO.input(BUTTON_PIN):
			self.begin_press()
		else:
			self.end_press()
			
	def begin_press(self):
		self.started = time.time()

	def end_press(self):
		self.ended = time.time()
		if not self.started:
			return

		if self.ended-self.started < RESETPRESSTIME:
			self.begin_shutdown()

	def begin_shutdown(self):
		subprocess.call(COMMAND.split())


GPIO.setmode(GPIO.BOARD)
GPIO.setup(PWR_PIN, GPIO.OUT, initial=GPIO.HIGH)
GPIO.setup(LED_PIN, GPIO.OUT, initial=GPIO.HIGH)
GPIO.setup(BUTTON_PIN, GPIO.IN)


pbh = PowerButtonHandler()
GPIO.add_event_detect(BUTTON_PIN, GPIO.BOTH, callback=pbh.press_event, bouncetime=BOUNCETIME)


while 1:
    time.sleep(0.1)

src/variants/lpfrg/filesystem/root/home/pi/scripts/factoryreset.py
#!/usr/bin/env python
# Author: Erik Heidstra <ErikHeidstra@live.nl>

import time
import RPi.GPIO as GPIO
import subprocess
import pygtk
import gtk, gobject, glib

FULLFACTORYRESETCOMMAND = r"echo 'full' | sudo tee /boot/factoryreset.txt && sudo sed -i 's@root=/dev/mmcblk0p2@root=/dev/mmcblk0p6@' /boot/cmdline.txt && sudo reboot"
FASTFACTORYRESETCOMMAND = r"echo 'fast' | sudo tee /boot/factoryreset.txt && sudo sed -i 's@root=/dev/mmcblk0p2@root=/dev/mmcblk0p6@' /boot/cmdline.txt && sudo reboot"

BUTTON_PIN = 16

BOUNCETIME = 100 # minimal press interval in ms
RESETPRESSTIME = 15 # duration of press before factory reset is initiated (in seconds)

RESPONSE_CANCEL=0
RESPONSE_FAST=1
RESPONSE_FULL=2

class PowerButtonHandler:
	def __init__(self):
		self.started = None
		self.ended = None
		self.confirmation_open = False

	def press_event(self, e):
		if GPIO.input(BUTTON_PIN):
			 # Interrupt the gtk main loop
			glib.idle_add(self.begin_press)
		else:
			# Interrupt the gtk main loop
			glib.idle_add(self.end_press)

	def begin_press(self):
		self.started = time.time()

	def end_press(self):
		self.ended = time.time()
		if not self.started:
			return

		if self.ended-self.started >= RESETPRESSTIME:
			self.confirm_factoryreset()

	def confirm_factoryreset(self):
		self.confirmation_open = True
		message = gtk.Dialog(buttons=("Cancel", RESPONSE_CANCEL, "Fast", RESPONSE_FAST, "Full", RESPONSE_FULL))
		message.set_default_size(400, 200)
		message.set_title("Factory reset")
		
		image = gtk.Image()
		image.set_from_stock(gtk.STOCK_DIALOG_WARNING, gtk.ICON_SIZE_DIALOG)

		label = gtk.Label("You are about to restore your printer to factory defaults. This will erase all your data. This action cannot be undone. You may choose for a fast restore or a full restore.\n\nWhich restore would you like to carry out?")
		label.set_line_wrap(True)

		content = message.get_content_area()
  		
  		box = gtk.HBox()
  		
		box.pack_start(image, True, True, 10)
		box.pack_start(label, True, True, 10)

		box.show()
		image.show()
		label.show()

		content.add(box)
		message.set_keep_above(True)
		response = message.run()
		self.confirmation_open = False
		message.destroy()
		
		if response == RESPONSE_FULL:
			subprocess.Popen(FULLFACTORYRESETCOMMAND, shell=True)
		elif response == RESPONSE_FAST:
			subprocess.Popen(FASTFACTORYRESETCOMMAND, shell=True)

pbh = PowerButtonHandler()
GPIO.setmode(GPIO.BOARD)
GPIO.setup(BUTTON_PIN, GPIO.IN)
GPIO.add_event_detect(BUTTON_PIN, GPIO.BOTH, callback=pbh.press_event, bouncetime=BOUNCETIME)

glib.threads_init() 
gtk.main()

src/variants/lpfrg/filesystem/root/home/pi/LpfrgScreensaver/lpfrgscreensaver.py
#!/usr/bin/python
from __future__ import division

import os
import sys

import pygtk
import gtk, gobject, cairo
from gtk import gdk
import ctypes as ct

import json, yaml, requests

import math
from random import randint
import RPi.GPIO as GPIO
import time

LED_PIN = 22 
PWR_PIN = 11 # Toggles power supply
BUTTON_PIN = 16

BOUNCETIME = 1000 # minimal press interval in ms

FONT_FACE = "/home/pi/LpfrgScreensaver/futura.ttf"
IMAGE_FOLDER = "/home/pi/LpfrgScreensaver/logos/"
OCTOPRINT_CONFIG_PATH = "/home/pi/.octoprint/config.yaml"
OCTOPRINT_MACHINE_DATA_PATH = "/home/pi/.octoprint/data/lui/machine.json"
JOB_URL = "http://localhost:5000/api/job?apikey={api_key}"
DEFAULT_MODEL = "bolt"
IDLE_STR = "IDLE"
FINISHED_STR = "PRINT FINISHED"

# the secret sauce is to get the "window id" out of $XSCREENSAVER_WINDOW
# code comes from these two places:
# 1) http://pastebin.com/nSCiq1P3
# 2) http://stackoverflow.com/questions/4598581/python-clutter-set-display

#To add: http://stackoverflow.com/questions/7016509/a-way-to-animate-transition-with-python-gtk-and-cairo
# https://cairographics.org/cookbook/animationrotation/



_initialized = False
def create_cairo_font_face_for_file (filename, faceindex=0, loadoptions=0):
    "given the name of a font file, and optional faceindex to pass to FT_New_Face" \
    " and loadoptions to pass to cairo_ft_font_face_create_for_ft_face, creates" \
    " a cairo.FontFace object that may be used to render text with that font."
    global _initialized
    global _freetype_so
    global _cairo_so
    global _ft_lib
    global _ft_destroy_key
    global _surface

    CAIRO_STATUS_SUCCESS = 0
    FT_Err_Ok = 0

    if not _initialized:
        # find shared objects
        _freetype_so = ct.CDLL("libfreetype.so.6")
        _cairo_so = ct.CDLL("libcairo.so.2")
        _cairo_so.cairo_ft_font_face_create_for_ft_face.restype = ct.c_void_p
        _cairo_so.cairo_ft_font_face_create_for_ft_face.argtypes = [ ct.c_void_p, ct.c_int ]
        _cairo_so.cairo_font_face_get_user_data.restype = ct.c_void_p
        _cairo_so.cairo_font_face_set_user_data.argtypes = (ct.c_void_p, ct.c_void_p, ct.c_void_p, ct.c_void_p)
        _cairo_so.cairo_set_font_face.argtypes = [ ct.c_void_p, ct.c_void_p ]
        _cairo_so.cairo_font_face_status.argtypes = [ ct.c_void_p ]
        _cairo_so.cairo_font_face_destroy.argtypes = (ct.c_void_p,)
        _cairo_so.cairo_status.argtypes = [ ct.c_void_p ]
        # initialize freetype
        _ft_lib = ct.c_void_p()
        status = _freetype_so.FT_Init_FreeType(ct.byref(_ft_lib))
        if  status != FT_Err_Ok :
            raise RuntimeError("Error %d initializing FreeType library." % status)
        #end if

        class PycairoContext(ct.Structure):
            _fields_ = \
                [
                    ("PyObject_HEAD", ct.c_byte * object.__basicsize__),
                    ("ctx", ct.c_void_p),
                    ("base", ct.c_void_p),
                ]
        #end PycairoContext

        _surface = cairo.ImageSurface(cairo.FORMAT_A8, 0, 0)
        _ft_destroy_key = ct.c_int() # dummy address
        _initialized = True
    #end if

    ft_face = ct.c_void_p()
    cr_face = None
    try :
        # load FreeType face
        status = _freetype_so.FT_New_Face(_ft_lib, filename.encode("utf-8"), faceindex, ct.byref(ft_face))
        if status != FT_Err_Ok :
            raise RuntimeError("Error %d creating FreeType font face for %s" % (status, filename))
        #end if

        # create Cairo font face for freetype face
        cr_face = _cairo_so.cairo_ft_font_face_create_for_ft_face(ft_face, loadoptions)
        status = _cairo_so.cairo_font_face_status(cr_face)
        if status != CAIRO_STATUS_SUCCESS :
            raise RuntimeError("Error %d creating cairo font face for %s" % (status, filename))
        #end if
        # Problem: Cairo doesn't know to call FT_Done_Face when its font_face object is
        # destroyed, so we have to do that for it, by attaching a cleanup callback to
        # the font_face. This only needs to be done once for each font face, while
        # cairo_ft_font_face_create_for_ft_face will return the same font_face if called
        # twice with the same FT Face.
        # The following check for whether the cleanup has been attached or not is
        # actually unnecessary in our situation, because each call to FT_New_Face
        # will return a new FT Face, but we include it here to show how to handle the
        # general case.
        if _cairo_so.cairo_font_face_get_user_data(cr_face, ct.byref(_ft_destroy_key)) == None :
            status = _cairo_so.cairo_font_face_set_user_data \
              (
                cr_face,
                ct.byref(_ft_destroy_key),
                ft_face,
                _freetype_so.FT_Done_Face
              )
            if status != CAIRO_STATUS_SUCCESS :
                raise RuntimeError("Error %d doing user_data dance for %s" % (status, filename))
            #end if
            ft_face = None # Cairo has stolen my reference
        #end if

        # set Cairo font face into Cairo context
        cairo_ctx = cairo.Context(_surface)
        cairo_t = PycairoContext.from_address(id(cairo_ctx)).ctx
        _cairo_so.cairo_set_font_face(cairo_t, cr_face)
        status = _cairo_so.cairo_font_face_status(cairo_t)
        if status != CAIRO_STATUS_SUCCESS :
            raise RuntimeError("Error %d creating cairo font face for %s" % (status, filename))
        #end if

    finally :
        _cairo_so.cairo_font_face_destroy(cr_face)
        _freetype_so.FT_Done_Face(ft_face)
    #end try

    # get back Cairo font face as a Python object
    face = cairo_ctx.get_font_face()
    return face
#end create_cairo_font_face_for_file

class Screen(gtk.DrawingArea):
    """ This class is a Drawing Area"""
    def __init__(self):
        super(Screen,self).__init__()
        ## Old fashioned way to connect expose. I don't savvy the gobject stuff.
        self.connect ( "expose_event", self.expose )
        self.connect ( "size-allocate", self.size )
        ## This is what gives the animation life!
        gobject.timeout_add( 50, self.tick ) # Go call tick every 50 whatsits.

    def tick ( self ):
        ## This invalidates the screen, causing the expose event to fire.
        self.alloc = self.get_allocation()
        rect = gtk.gdk.Rectangle(self.alloc.x, self.alloc.y, self.alloc.width, self.alloc.height)
        self.window.invalidate_rect(rect, True)
        return True # Causes timeout to tick again.

    ## When expose event fires, this is run
    def expose(self, widget, event):
        self.cr = self.window.cairo_create()
        self.draw()

class ProgressDisplay(Screen):
    """This class is also a Drawing Area, coming from Screen."""
    def __init__ ( self ):
        Screen.__init__( self )
        self.octoprint_comm = OctoPrintComm()
        self.progress_string = IDLE_STR
        self.image = None
        self.counter = 0
    
    def size(self, widget, event):
        self.screen_w, self.screen_h = self.window.get_size()

        image_path = self.octoprint_comm.get_image_file()

        if image_path:
            self.image = cairo.ImageSurface.create_from_png(image_path)
            self.image_w = self.image.get_width()
            self.image_h = self.image.get_height()
        else:
            self.image_w = 0
            self.image_h = 0

        self.font_face = create_cairo_font_face_for_file(FONT_FACE, 0)
        
        self.counter = 0
        self.fade_duration = 350

        self.text_margin = 20

        self.image_x = (self.screen_w - self.image_w) / 2
        self.image_y = randint(0, self.screen_h - self.image_h - self.text_margin - 50)

        self.progress_string = self.octoprint_comm.get_progress_string()
       
    def draw( self ):
        ## A shortcut
        self.cr.set_source_rgba(0.0, 0.0, 0.0, 1)
        self.cr.rectangle(0, 0, self.screen_w, self.screen_h)
        self.cr.fill()

        if self.counter < self.fade_duration:
            self.drawProgress(self.cr, self.counter / self.fade_duration)
            if self.image:
                self.drawImage(self.cr, self.counter / self.fade_duration)
            
        
        self.counter += 1

        if self.counter >= self.fade_duration:
            self.progress_string = self.octoprint_comm.get_progress_string()
            self.image_y = randint(0, self.screen_h - self.image_h - self.text_margin - 50)
            self.counter = 0
       
    def drawProgress(self, cr, progress):
        cr.set_source_rgba(1, 1, 1, self.getAlpha(progress))
        cr.set_font_face(self.font_face)
        cr.set_font_size(45)
        
        (x, y, width, height, dx, dy) = cr.text_extents(self.progress_string)
        cr.move_to(self.image_x + (self.image_w - width)/2, self.image_y + self.image_h + height + self.text_margin)    
        cr.show_text(self.progress_string)


    def drawImage ( self, cr, progress):
        cr.set_source_surface(self.image, self.image_x, self.image_y)
        cr.rectangle( self.image_x, self.image_y, self.image_w, self.image_h )
        cr.clip()
        alpha = self.getAlpha(progress)
        cr.paint_with_alpha(alpha)

    def getAlpha(self, progress):
        if progress <= 0.2:
            return 5*progress
        elif progress <= 0.8:
            return 1
        else:
            return 5 * (1-progress)

class ScreenSaverWindow(gtk.Window):

    def __init__(self):
        gtk.Window.__init__(self)
        self.screen_w = 600
        self.screen_h = 1024

    def realize(self):
        if self.flags() & gtk.REALIZED:
            return

        ident = os.environ.get('XSCREENSAVER_WINDOW')
        if not ident is None:
            self.window = gtk.gdk.window_foreign_new(int(ident, 16))
            self.window.set_events (gdk.EXPOSURE_MASK | gdk.STRUCTURE_MASK)
            
            x, y, w, h, depth = self.window.get_geometry()
            self.size_allocate(gtk.gdk.Rectangle(x, y, w, h))
            self.set_default_size(w, h)
            self.set_decorated(False)

            self.window.set_user_data(self)
            self.style.attach(self.window)

        if self.window == None:
            self.window = gdk.Window(None, self.screen_w, self.screen_h, gdk.WINDOW_TOPLEVEL,
                                     (gdk.EXPOSURE_MASK | gdk.STRUCTURE_MASK),
                                     gdk.INPUT_OUTPUT)
 
            self.set_default_size(self.screen_w, self.screen_h)
            self.set_decorated(False)   
            self.window.set_user_data(self)       
            self.style.attach(self.window)  

        if self.window != None:
            self.set_flags(self.flags() | gtk.REALIZED)

class OctoPrintComm():
    def __init__(self):
        self.config_data = None
        self.machine_data = None
        self.api_key = self._read_api_key()
        self.model = self._read_model()
        self.job_url = JOB_URL.format(api_key=self.api_key)

    # Public methods
    def get_progress_string(self):
        job = self._get_job_data()

        if job and "state" in job and "progress" in job:
            if job["state"] == "Operational" and job["progress"]["completion"] == 100:
                return FINISHED_STR
            elif job["state"] == "Printing":
                return "{0:.0f}%".format(job["progress"]["completion"])
        
        return IDLE_STR

    def get_image_file(self):
        if self.model:
            path = os.path.join(IMAGE_FOLDER, self.model + ".png")
            if os.path.exists(path):
                return path

    # Private methods
    def _read_config(self):
        if os.path.exists(OCTOPRINT_CONFIG_PATH):
            with open(OCTOPRINT_CONFIG_PATH, "rb") as fp:
                self.config_data = yaml.safe_load(fp)
    
    def _read_machine_data(self):
        if os.path.exists(OCTOPRINT_MACHINE_DATA_PATH):
            with open(OCTOPRINT_MACHINE_DATA_PATH, "rb") as fp:
                self.machine_data = json.load(fp)

    def _read_api_key(self):
        if not self.config_data:
            self._read_config()

        if self.config_data and "api" in self.config_data and "key" in self.config_data["api"]:
            return self.config_data["api"]["key"]

    def _read_model(self):
        if not self.machine_data:
            self._read_machine_data()

        if self.machine_data and "_default" in self.machine_data:
            for _, item in self.machine_data["_default"].iteritems():
                if "property" in item and "value" in item and item["property"] == "machine_type" and item["value"]:
                        return item["value"].lower()
        
        return DEFAULT_MODEL

    def _get_job_data(self):
        try:
            response = requests.get(self.job_url)
        except:
            return None
        
        if response.status_code == 200:
            try:
                return response.json()
            except:
                return None
       
        

window = ScreenSaverWindow()
window.set_title('Floaters')
window.set_default_size(600, 1024)
window.realize()

window.modify_bg(gtk.STATE_NORMAL, gdk.color_parse("black"))

window.connect( "delete-event", gtk.main_quit )
widget = ProgressDisplay()
widget.show()
window.add(widget)
window.present()

def shutdown_button_press(c):
    import subprocess
    subprocess.call("xscreensaver-command -deactivate".split())
    
GPIO.setmode(GPIO.BOARD)
GPIO.setup(BUTTON_PIN, GPIO.IN)
GPIO.add_event_detect(BUTTON_PIN, GPIO.RISING, callback=shutdown_button_press, bouncetime=BOUNCETIME)

gtk.main()
